rpc_state_reader_sn_api/src/lib.rs

```diff
@@ -0,0 +1,994 @@
+        }
+
+        Ok(RpcCallInfo {
+            execution_resources,
+            retdata,
+            calldata,
+            internal_calls,
+        })
+    }
+}
+
+impl<'de> Deserialize<'de> for TransactionTrace {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let value: serde_json::Value = Deserialize::deserialize(deserializer)?;
+
+        let validate_invocation = value["validate_invocation"].clone();
+        let function_invocation = value["function_invocation"].clone();
+        let fee_transfer_invocation = value["fee_transfer_invocation"].clone();
+        let signature_value = value["signature"].clone();
+
+        Ok(TransactionTrace {
+            validate_invocation: serde_json::from_value(validate_invocation)
+                .map_err(serde::de::Error::custom)?,
+            function_invocation: serde_json::from_value(function_invocation)
+                .map_err(serde::de::Error::custom)?,
+            fee_transfer_invocation: serde_json::from_value(fee_transfer_invocation)
+                .map_err(serde::de::Error::custom)?,
+            signature: serde_json::from_value(signature_value).map_err(serde::de::Error::custom)?,
+        })
+    }
+}
+
+impl RpcState {
+    /// Requests the transaction trace to the Feeder Gateway API.
+    /// It's useful for testing the transaction outputs like:
+    /// - execution resources
+    /// - actual fee
+    /// - events
+    /// - return data
+    pub fn get_transaction_trace(&self, hash: &TransactionHash) -> TransactionTrace {
+        let chain_name = self.get_chain_name();
+        let response = ureq::get(&format!(
+            "https://{}.starknet.io/feeder_gateway/get_transaction_trace",
+            chain_name
+        ))
+        .query("transactionHash", &hash.0.to_string())
+        .call()
+        .unwrap();
+
+        serde_json::from_str(&response.into_string().unwrap()).unwrap()
+    }
+
+    /// Requests the given transaction to the Feeder Gateway API.
+    pub fn get_transaction(&self, hash: &TransactionHash) -> Transaction {
+        let params = ureq::json!({
+            "jsonrpc": "2.0",
+            "method": "starknet_getTransactionByHash",
+            "params": [hash.to_string()],
+            "id": 1
+        });
+        let result = self.rpc_call::<serde_json::Value>(&params).unwrap()["result"].clone();
+
+        match result["type"].as_str().unwrap() {
+            "INVOKE" => match result["version"].as_str().unwrap() {
+                "0x0" => Transaction::Invoke(InvokeTransaction::V0(
+                    serde_json::from_value(result).unwrap(),
+                )),
+                "0x1" => Transaction::Invoke(InvokeTransaction::V1(
+                    serde_json::from_value(result).unwrap(),
+                )),
+                _ => unreachable!(),
+            },
+            _ => unreachable!(),
```

Comment:
I don't really like using `unreachable!()` with externally provided inputs since there's no guarantee that they'll be really unreachable. How about using a `panic!("Unexpected value returned from the API.")`?

Reply:
Sorry these should be `unimplemented!()` as for now we're only handling invoke transactions. Thanks!